### ISSUE TITLE
Random val_loss behaviour

### ISSUE THREAD

**Issue by victor-roris** [divamgupta/image-segmentation-keras#201]
Hi, 
I am a newbie in this ML area. 
I am trying to use this tool to identify different segments or parts of a document (header, text, title, table, figure, ...). I manually annotated 3000 examples.  

I used the following code to train:

```
model.train(
    train_images =  keimseg_input_folder + "train_images",
    train_annotations = keimseg_input_folder + "train_segmentation",
    checkpoints_path = keimseg_checkpoints_name , 
    epochs=5,
    steps_per_epoch=512,
    batch_size=4,
    validate=True,
    val_images = keimseg_input_folder + "val_images" , 
    val_annotations = keimseg_input_folder + "val_segmentation"
)
``` 

I tried different models (fcn_8 , fcn_32 , fcn_8_vgg, pspnet, vgg_unet, ...) and I always obtained a similar behaviour. The val_loss ascend and descent randomly, instead of the expected progressive descent. For example:

![Captura de pantalla -2020-04-28 15-52-13](https://user-images.githubusercontent.com/33596234/80514181-75006200-8980-11ea-8024-0ab0e4c602ce.png) 

1) Am I doing something wrong (bad hyperparameters, bigger dataset, etc)? 
2) What is the result more important here, the `val_loss` or `val_accuracy`?
3) What can be the best model for my goal (currently, I have the best performance in fcn_8 and resnet50_segnet)?

Thanks in advance,

**Comment by divamgupta** [comment]
The val_loss printed is not on the whole dataset, but a small number of mini-batches. 
You should use the evaluate function to get the IoU scores on the complete test/val data.
Try finetuning a pre-trained PSPNet model.

**Comment by victor-roris** [comment]
Thanks @divamgupta for your response.
Finetuning a pre-trained PSPNet model gives me a better accuracy:

![Captura de pantalla -2020-04-30 10-18-46](https://user-images.githubusercontent.com/33596234/80688431-0d572d80-8acc-11ea-8e6b-1f2560e250ed.png)

If val_loss is not representative of the complete validation data, there are some easy way to know which is the appropriate number of epochs?

**Comment by divamgupta** [comment]
One thing you can do is setting val_steps_per_epoch as the number of images divided by val_batch_size.

**Comment by victor-roris** [comment]
Thanks @divamgupta !
It makes a lot of sense. I didn't see the fields `val_steps_per_epoch` and `val_batch_size`. If it helps someone, my final code is:
```
train_batch_size = 4
steps_per_epoch = len(trn_dataset) // train_batch_size

val_batch_size = 4
val_steps_per_epoch = len(val_dataset) // val_batch_size

model.train(
    train_images =  keimseg_input_folder + "train_images",
    train_annotations = keimseg_input_folder + "train_segmentation",
    checkpoints_path = keimseg_checkpoints_name , 
    epochs=5,
    steps_per_epoch=steps_per_epoch,
    batch_size=train_batch_size,
    validate=True,
    val_steps_per_epoch=val_steps_per_epoch,
    val_batch_size=val_batch_size,
    val_images = keimseg_input_folder + "val_images" , 
    val_annotations = keimseg_input_folder + "val_segmentation"
)
```

In any case, I still having problems with the `val_loss` (it ascends and descents randomly), but I tested with other libraries and the behaviour of the `loss` and `accuracy` is pretty similar. I suppose its a problem with my own dataset. 

Thanks again for the help and for your fantastic library!

**Comment by varungupta31** [comment]
Hey @victor-roris, can you please share your tensorflow version?
(I'm using older tf version due to GPU constraints, and running into some issues while training this model)
